src/internal/queues.mjs

```diff
@@ -44,19 +29,37 @@ export class Queue {
   }
 }
 
-export class Exchange extends Queue {
+class Consumer {
+  constructor (queueItem) {
+    this.queueItem = queueItem
+  }
+  isEmpty () {
+    return !this.queueItem.previous
+  }
   shift () {
-    throw new Error('Unsupported')
+    if (this.isEmpty()) throw new Error('Queue is empty')
+    const data = this.queueItem.previous.data
+    this.queueItem = this.queueItem.previous
+    return data
+  }
+  clone () {
+    return new Consumer(this.queueItem)
+  }
+}
+
+export class Exchange {
+  constructor (queueItem) {
+    this.tail = new QueueItem() // an empty queue points to a tail node
   }
 
-  spawnConsumer () {
-    if (!this.head) throw new Error('You cannot spawn a new consumer after setting calling noMoreConsumers')
-    return new Consumer(this.head)
+  push (data) {
+    const newItem = new QueueItem(data)
+    this.tail.previous = newItem
+    this.tail = newItem
   }
 
-  noMoreConsumers () {
```

Comment:
I still rather liked `noMoreConsumers`. Sure it was a bit cryptic and long as a name, but I think that's exactly what you want for a method that begs to reader to go look carefully at what is happening. It also created a spot where detailed documentation could be coupled to the smallest chunk of code that wholly encapsulated the desired behavior.

Reply:
The current design makes *noMoreConsumers* redundant and I am pretty happy with it.
It's not clear what you ask here